R/parallel_new.R

```diff
@@ -0,0 +1,205 @@
+# This are new changes for parallel process via future _or_ mirai. These
+# functions should be able to replace what is contained in `parallel.R`
+
+# ------------------------------------------------------------------------------
+
+has_non_par_pkgs <- function(object, control, verbose = FALSE) {
+  pkgs <- character(0)
+
+  if (!is.null(object)) {
+    pkgs <- required_pkgs(object)
+  }
+  if (!is.null(control)) {
+    pkgs <- c(pkgs, control$pkgs)
+  }
+  pkgs <- unique(pkgs)
+  if (length(pkgs) == 0) {
+    return(FALSE)
+  }
+  naughty_list <- c("keras", "rJava")
+  has_pkg <- pkgs %in% naughty_list
+  if (any(has_pkg)) {
+    pkgs <- pkgs[has_pkg]
+    if (verbose) {
+      cli::cli_inform(
+        "These packages cannot be used with explicit parallel processing: {.pkg {pkgs}}."
+      )
+    }
+  }
+  any(has_pkg)
+}
+
+# ------------------------------------------------------------------------------
+
+future_avail <- function() {
+  any(search() == "package:future")
+}
+mirai_avail <- function() {
+  any(search() == "package:mirai")
+}
```

Comment:
It's possible to set daemons without loading the mirai package, and same to plan without future loaded. Could the logic here just be based on the information in `future::nbrOfWorkers()` and `mirai::status()`? Checking whether the packages are loaded feels like an unusual mechanism.

Reply:
The plan is for both mirai and future to be in suggests so someone will have to install both even if they don't want to run in parallel.